google/cloud/storage/client.h

```diff
@@ -2162,10 +2162,18 @@ class Client {
    * @param object_name the name of the object, note that the object may not
    *     exist for signed URLs that upload new objects.
    * @param options a list of optional parameters for the signed URL, this
-   *     include: `ExpirationTime`, `MD5HashValue`, `ContentType`, and
-   *     `AddExtensionHeaderOption`. The `AddExtensionHeader()` function
-   *     provides a simpler way to create extension headers. Note that you can
-   *     provides multiple values of this option.
+   *     include: `ExpirationTime`, `MD5HashValue`, `ContentType`,
+   *     `AddExtensionHeaderOption`, and `AddQueryParameterOption`. The
+   *     `AddExtensionHeader()` function provides a simpler way to create
+   *     extension headers. Note that you can provides multiple values of this
+   *     option. Likewise, the following helper functions can create properly
+   *     formatted query parameters: `WithAcl()`, `WithBilling()`,
+   *     `WithCompose()`, `WithCors()`, `WithDelimiter()`,
+   *     `WithEncryption()`, `WithEncryptionConfig()`,`WithGeneration()`,
+   *     `WithGenerationMarker()`, `WithLifecycle()`, `WithLocation()`,
+   *     `WithLogging()`, `WithMarker()`, `WithMaxKeys()`, `WithPrefix()`,
+   *     `WithResponseContentDisposition()`, `WithResponseContentType()`,
+   *     `WithStorageClass()`, `WithTagging()`, `WithUserProject()`.
```

Comment:
[GCS documentation](https://cloud.google.com/storage/docs/access-control/signed-urls#about-canonical-resources) states that prefix, max-keys, and delimiter should not be included.

Additionally, I think you can only include one subresource (acl, billing, encryption). When I try multiple, the API responds with the first in line. Though this isn't specifically documented.

Reply:
What if, rather than implementing a separate `Option` subclass for each subresource, we had a more generic `WithSubresource()` option? This could either accept a string (which would have the downside of allowing users to specify disallowed strings like `"prefix"`) or something that gets converted to the appropriate string, e.g. an enum class, which might look a little long-winded: `WithSubresource(Subresources::CORS)`.

This would also allow us to add a note saying that if multiple `WithSubresource()` options are specified, we'll only use the last one.

Thoughts?

Reply:
I am intentionally being lazy about this, and would prefer not to implement `WithSubresource`, we will quickly have to implement a full new API if we do this.  For example, the subresources really only apply to `GET` requests for a bucket, should we create a function called `CreateV2SignedUrlGetBucket` (and its corresponding request) and only make them applicable to that?  Some of the other options only apply to `HEAD` requests on an object.  So maybe we should have `CreateV2SignedUrlHeadObject`?  But there are clearly several of these functions, maybe we should create a `V2Signer` class that has `GetBucketMetadata`, `GetObjectMetadata` functions each one of which signs a different thing?

Now we have a second "gcs::Client" that does not perform the action, only creates the URL to perform the action over the XML API.  I do not think we should implement that, not for a number of relatively low use cases.

The API to sign a "GET object" and "PUT object" and "GET bucket" are there, they are easy to use, the rest are possible.